service/routes.py

```diff
@@ -172,11 +172,34 @@ def get_customer_addresses(customer_id):
 def list_customers():
     """Returns all of the customers"""
     app.logger.info("Request for customer list")
-    category = request.args.get("category")
-    name = request.args.get("first_name")
-
-    results = [customer.serialize() for customer in Customer.all()]
+    
+    all_query_key = ["username", "first_name", "last_name", "addresses"]
+    for key in request.args.keys():
+      if key not in all_query_key:
+        message = {
+            "error": "Conflict",
+            "message": "The query key: '" + key + "' is not supported."
+            }
 
+        return make_response(
+            jsonify(message), status.HTTP_409_CONFLICT
+        ) 
+    
+    username = request.args.get("username")
+    first_name = request.args.get("first_name")
+    last_name = request.args.get("last_name")
+    #addresses = request.args.get("addresses")
+    if username:
+      customers = Customer.find_by_name(username)
+    elif first_name:
+      customers = Customer.find_by_first_name(first_name)
+    elif last_name:
+      customers = Customer.find_by_last_name(last_name)
+    #elif addresses:
+    #  customers = Customer.find_by_addresses(addresses)
```

Comment:
I think it's fine to not allow querying by addresses here, but I guess we should remove addresses from all_query_key on line 176 if this is the case

Reply:
Fixed. I already removed it.